### PR TITLE
#67 - Project Model

### PR DESCRIPTION
**Pull request by uU-Uuu**

### CODE REVIEW

**Review by SaberanPWNZ — APPROVED** [review]

**Review by PolinaCh14 — APPROVED** [review]

File: projects/tests.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        project_upd.save()
+        self.assertEqual(len(self.project.history.all()), 2)
+
+    def test_subscription(self):
+        """test assigning investor to a project"""
+        self.project.investor.add(self.investor_)
+        self.assertTrue(Subscription.objects.filter(
+            project_id=self.project.project_id,
+            investor_id=self.investor_.id).exists(),
+            True)
```
Comment (mehalyna) [inline projects/tests.py:88]: The `assertTrue` assertion should not have a second argument like `True`.

File: projects/tests.py
Comment (mehalyna) [inline projects/tests.py:1]: It's good practice to test the edge cases of allowed values, especially for fields with validation constraints like `risk` (0.0 to 1.0).

Comment (mehalyna) [inline projects/tests.py:1]: Although you’ve added a field for `media_files`, the tests don’t seem to cover it. You should test that the field accepts valid binary data and raises errors for invalid data types.

Comment (mehalyna) [inline projects/tests.py:1]: As the `clean_duration` method checks for positive values in the duration field, add a test to ensure it works as expected.

Comment (mehalyna) [inline projects/tests.py:1]: You should test that a project can be created even if the `business_plan` URL is not provided, as it’s an optional field

File: projects/models.py [hunk 2]
```diff
@@ -1,3 +1,89 @@
...
+    investor = models.ManyToManyField(
+        InvestorProfile, related_name='projects', blank=True, through="Subscription")
+    title = models.CharField(max_length=255)
+    risk = models.FloatField(
+        validators=[MinValueValidator(0.0), MaxValueValidator(1.0)])
+    description = models.TextField()
+    business_plan = models.URLField(blank=True, null=True)
+    amount = models.DecimalField(
+        max_digits=10, decimal_places=2, validators=[MinValueValidator(0.0)])
+    media_files = models.BinaryField(blank=True, null=True)
```
[the remark this answers is not in the document]
Comment (uU-Uuu) [reply projects/models.py]: done

File: projects/models.py [hunk 3]
```diff
@@ -1,3 +1,89 @@
...
+        CLOSED = 4
+
+    project_id = models.UUIDField(
+        primary_key=True, default=uuid4, editable=False)
+    startup = models.ForeignKey(
+        StartUpProfile, related_name='projects', on_delete=models.CASCADE)
+    investor = models.ManyToManyField(
+        InvestorProfile, related_name='projects', blank=True, through="Subscription")
+    title = models.CharField(max_length=255)
+    risk = models.FloatField(
```
[the remark this answers is not in the document]
Comment (uU-Uuu) [reply projects/models.py:51]: done

File: projects/models.py [hunk 4]
```diff
@@ -1,3 +1,89 @@
...
+        validators=[MinValueValidator(0.0), MaxValueValidator(1.0)])
+    description = models.TextField()
+    business_plan = models.URLField(blank=True, null=True)
+    amount = models.DecimalField(
+        max_digits=10, decimal_places=2, validators=[MinValueValidator(0.0)])
+    media_files = models.BinaryField(blank=True, null=True)
+    status = models.IntegerField(
```
[the remark this answers is not in the document]
Comment (uU-Uuu) [reply projects/models.py:61]: done

File: projects/models.py
[the remark this answers is not in the document]
Comment (uU-Uuu) [reply projects/models.py:1]: done

File: projects/models.py [hunk 5]
```diff
@@ -1,3 +1,89 @@
...
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    history = HistoricalRecords()
+
+    def __str__(self):
+        return self.title
+
+    def clean_duration(self):
+        if self.duration is not None and self.duration <= timedelta(0):
+            raise ValidationError('Duration must be a positive value')
```
[the remark this answers is not in the document]
Comment (uU-Uuu) [reply projects/models.py:108]: done

File: projects/tests.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    def tearDown(self):
+        self.project.delete()
```
[the remark this answers is not in the document]
Comment (uU-Uuu) [reply projects/tests.py]: done

File: projects/tests.py (hunk 1 above)
Comment (uU-Uuu) [reply projects/tests.py]: done

File: projects/tests.py
Comment (uU-Uuu) [reply projects/tests.py:1]: done

Comment (uU-Uuu) [reply projects/tests.py:1]: done

Comment (uU-Uuu) [reply projects/tests.py:1]: done

Comment (uU-Uuu) [reply projects/tests.py:1]: done
